### ISSUE TITLE
Illegal instruction issue

### ISSUE THREAD

**Issue by nadendlaprasad** [ShiftMediaProject/gnutls#12]
gnutls x64 build throws `Illegal Instruction  ` while executing  `sha1_block_data_order(&octx, data, t2);`  function which is inside  **lib\accelerated\x86\sha-x86-ssse3.c file**. It works well with x86 build.

could you please let me know where can i get the x64 version of this file or is this file meant for both x86 and x64 architectures ?

**Comment by Sibras** [comment]
sha1_block_data_order requires that your cpu support a minimum of ssse3. If it does not then the function will not execute. What cpu are you running on? If it is an older one then it may not have the required feature support.

**Comment by nadendlaprasad** [comment]
I am using intel core i7-2640M , x64 based processor

**Comment by nadendlaprasad** [comment]
I have listed out my CPU supported instruction set using **coreinfo** tool from windows. My CPU does support SSSE3( Supports Supplemental SIMD Extensions 3) instruction

**Comment by nadendlaprasad** [comment]
It is executing the sha1_block_data_order(&octx, data, t2); function with x86  VS2017 build with out any issues. Facing this issue specifically with x64 build

**Comment by Sibras** [comment]
There are 2 different versions of sha1_block_data_order; 1 for x86 and 1 for x64. Clearly there is an issue with the x64 version. This would require using dissassembly view when debugging to step through the sha1_block_data_order function to work out which instruction is "invalid". If you have some small test code that causes the fault then post it so i can debug it.

**Comment by nadendlaprasad** [comment]
I am using **opendht**  `dht::crypto::generateIdentity() ` function to simulate this issue. Please do let me know if you are familiar with this. If not I'll share a sample program with you soon

**Comment by Sibras** [comment]
im not familiar with openhdt. If you provide a very simple (i.e the smallest amount of code needed to reproduce the issue) sample ill look over it.

**Comment by nadendlaprasad** [comment]
Here the sample code to reproduce the error

```
#include <iostream>
#include <gnutls/gnutls.h>
#include <gnutls/x509.h>
#include <algorithm>
#include<gnutls/abstract.h>
int main()
{
	gnutls_x509_privkey_t key;
	unsigned int key_length = 4096;
	if (gnutls_x509_privkey_init(&key) != GNUTLS_E_SUCCESS)
		std::cout << "Failed to create private key" << std::endl;
	else
		std::cout << key << std::endl;
	int err = gnutls_x509_privkey_generate(key, GNUTLS_PK_RSA, key_length, 0);
	if (err != GNUTLS_E_SUCCESS) {
		gnutls_x509_privkey_deinit(key);
		std::cout << "Can't generate RSA key pair: " << gnutls_strerror(err) << std::endl;
	}
	else {
		gnutls_privkey_t k{};
		gnutls_privkey_init(&k);
		if (gnutls_privkey_import_x509(k, key, GNUTLS_PRIVKEY_IMPORT_COPY) != GNUTLS_E_SUCCESS) {
			key = nullptr;
		}
gnutls_pubkey_t pk;
		gnutls_pubkey_init(&pk);
		if (gnutls_pubkey_import_privkey(pk, k, GNUTLS_KEY_KEY_CERT_SIGN | GNUTLS_KEY_CRL_SIGN, 0) != GNUTLS_E_SUCCESS)
			return {};
		unsigned char id[20];
		size_t sz = 20;
		const int flags = (20 == 32) ? GNUTLS_KEYID_USE_SHA256 : 0;



// This  function call invokes `sha1_block_data_order`	
          gnutls_pubkey_get_key_id(pk, flags, id, &sz);	
		
	}
	system("pause");
	return 0;
}


```
`gnutls_pubkey_get_key_id(pk, flags, id, &sz);	`   This is the function which invokes  `sha1_block_data_order` internally

**Comment by nadendlaprasad** [comment]
Facing the same issue with `sha256_block_data_order(&octx, data, t2);` function. could you please look into it

**Comment by Sibras** [comment]
Is this issue occurring with the latest version of the code in the repo (you may have to do a force update) as this should have been fixed with the error in sha1_block_data_order.

**Comment by nadendlaprasad** [comment]
yes, this issue is occurring with the latest version of the code

**Comment by Sibras** [comment]
OK should be fixed now. Just do a force update.
GnuTLS was using some really non obvious assembly generation from code taken directly from a really old OpenSSL commit. The last error was due to GnuTLS generating 1 file different to all the others due to an non compatible cpuid function.
Luckily they have fixed all this in the upcomming 3.6.x branch so this shouldnt be a problem with future versions either.

**Comment by nadendlaprasad** [comment]
Awesome. Great!
